ta/pkcs11/src/pkcs11_token.c

```diff
@@ -973,11 +973,57 @@ static enum pkcs11_rc set_pin(struct pkcs11_session *session,
 		return PKCS11_CKR_SESSION_READ_ONLY;
 
 	if (IS_ENABLED(CFG_PKCS11_TA_AUTH_TEE_IDENTITY) &&
-	    token->db_main->flags & PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH) {
-		rc = setup_identity_auth_from_pin(token, user_type, new_pin,
-						  new_pin_size);
-		if (rc)
+	    (token->db_main->flags &
+	     PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH)) {
+		rc = setup_identity_auth_from_pin(token, user_type,
+						  new_pin, new_pin_size);
+		if (rc == PKCS11_CKR_OK) {
+			goto update_db;
+		} else if (rc == PKCS11_CKR_PIN_INVALID &&
+			   !(token->db_main->flags &
+		      PKCS11_CKFT_USER_PIN_INITIALIZED)) {
+			/*
+			 * PIN was not compatible with TEE Identity
+			 * Authentication syntax so let's assume it might be a
+			 * new SO PIN to switch the authentication mode.
+			 */
+
+			/* Update mode flag if all PIN checks pass */
+			flags_clear |=
+				PKCS11_CKFT_PROTECTED_AUTHENTICATION_PATH;
+		} else {
+			return rc;
+		}
+	} else if ((user_type == PKCS11_CKU_SO) && !new_pin &&
+		   !(token->db_main->flags &
+		     PKCS11_CKFT_USER_PIN_INITIALIZED)) {
```

Comment:
Shouldn't this block also be covered by `IS_ENABLED(CFG_PKCS11_TA_AUTH_TEE_IDENTITY)`?

Reply:
Yes it should.